### PR TITLE
AArch64: Select FCANONICALIZE

### PR DESCRIPTION
**Pull request by wzssyqa**
FMINNM/FMAXNM instructions of AArch64 follow IEEE754-2008. We can use them to canonicalize a floating point number. And FMINNUM_IEEE/FMAXNUM_IEEE is used by something like expanding FMINIMUMNUM/FMAXIMUMNUM, so let's define them.

### CODE REVIEW

**Comment by wzssyqa** [comment]
ping

File: llvm/lib/Target/AArch64/AArch64InstrInfo.td [hunk 1]
```diff
@@ -5051,6 +5051,18 @@ def : Pat<(v1f64 (fmaxnum (v1f64 FPR64:$Rn), (v1f64 FPR64:$Rm))),
...
+def : Pat<(fminnum_ieee (f32 FPR32:$a), (f32 FPR32:$b)),
+          (FMINNMSrr FPR32:$a, FPR32:$b)>;
+def : Pat<(fmaxnum_ieee (f64 FPR64:$a), (f64 FPR64:$b)),
+          (FMAXNMDrr FPR64:$a, FPR64:$b)>;
+def : Pat<(fmaxnum_ieee (f32 FPR32:$a), (f32 FPR32:$b)),
+          (FMAXNMSrr FPR32:$a, FPR32:$b)>;
+def : Pat<(f32 (fcanonicalize f32:$a)),
+          (FMINNMSrr   f32:$a, f32:$a)>;
+def : Pat<(f64 (fcanonicalize f64:$a)),
+          (FMINNMDrr   f64:$a, f64:$a)>;
```
Comment (arsenm) [inline llvm/lib/Target/AArch64/AArch64InstrInfo.td:5065]: This should be 2 separate patches, one for canonicalize and one for fminnum_ieee. This is also missing the half and vector handling.

This is also missing test coverage, especially for canonicalize

File: llvm/test/CodeGen/AArch64/fp-fcanonicalize.ll [hunk 2]
```diff
@@ -0,0 +1,168 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py
+; RUN: llc --mtriple=aarch64 --mattr=+fullfp16 < %s | FileCheck %s --check-prefix=AARCH64
+
+declare half @llvm.fcanonicalize.f16(half)
+declare float @llvm.fcanonicalize.f32(float)
+declare double @llvm.fcanonicalize.f64(double)
+
+define half @fcanonicalize_half(half %x) {
```
Comment (arsenm) [inline llvm/test/CodeGen/AArch64/fp-fcanonicalize.ll:8]:
```suggestion
define half @fcanonicalize_f16(half %x) {
```

File: llvm/test/CodeGen/AArch64/fp-fcanonicalize.ll [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+; AARCH64-LABEL: fcanonicalize_half:
+; AARCH64:       // %bb.0:
+; AARCH64-NEXT:    fminnm h0, h0, h0
+; AARCH64-NEXT:    ret
+  %z = call half @llvm.canonicalize.f16(half %x)
+  ret half %z
+}
+
+define half @fcanonicalize_half_nnan(half %x) {
```
Comment (arsenm) [inline llvm/test/CodeGen/AArch64/fp-fcanonicalize.ll:17]:
```suggestion
define half @fcanonicalize_f16_nnan(half %x) {
```

File: llvm/test/CodeGen/AArch64/fp-fcanonicalize.ll [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+define <8 x half> @fcanonicalize_v8f16_nnan(<8 x half> %x) {
+; AARCH64-LABEL: fcanonicalize_v8f16_nnan:
+; AARCH64:       // %bb.0:
+; AARCH64-NEXT:    fminnm v0.8h, v0.8h, v0.8h
+; AARCH64-NEXT:    ret
+  %z = call nnan <8 x half> @llvm.canonicalize.v8f16(<8 x half> %x)
+  ret <8 x half> %z
+}
+
+define float @fcanonicalize_float(float %x) {
```
Comment (arsenm) [inline llvm/test/CodeGen/AArch64/fp-fcanonicalize.ll:80]:
```suggestion
define float @fcanonicalize_f32(float %x) {
```

**Review by arsenm — APPROVED** [review]
In future could try to select to a copy in cases where the source is known canonicalized already

File: llvm/lib/Target/AArch64/AArch64InstrInfo.td [hunk 5]
```diff
@@ -5063,6 +5063,12 @@ def : Pat<(fmaxnum_ieee (f32 FPR32:$a), (f32 FPR32:$b)),
 def : Pat<(fmaxnum_ieee (f16 FPR16:$a), (f16 FPR16:$b)),
           (FMAXNMHrr FPR16:$a, FPR16:$b)>;
 
+def : Pat<(f16 (fcanonicalize f16:$a)),
+          (FMINNMHrr   f16:$a, f16:$a)>;
```
Comment (davemgreen) [inline llvm/lib/Target/AArch64/AArch64InstrInfo.td:5067]: FP16 patterns require a hasFullFP16 predicate, otherwise the instructions will not be available. Same for vector patterns.

File: llvm/test/CodeGen/AArch64/fp-fcanonicalize.ll [hunk 6]
```diff
@@ -0,0 +1,168 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py
+; RUN: llc --mtriple=aarch64 --mattr=+fullfp16 < %s | FileCheck %s --check-prefix=AARCH64
```
Comment (davemgreen) [inline llvm/test/CodeGen/AArch64/fp-fcanonicalize.ll:2]: Remove --check-prefix=AARCH64, but it is worth adding a non-fp16 run line too. Something like
```
; RUN: llc -mtriple=aarch64 -verify-machineinstrs %s -o - | FileCheck %s --check-prefixes=CHECK,CHECK-NOFP16
; RUN: llc -mtriple=aarch64 -mattr=+fullfp16 -verify-machineinstrs %s -o - | FileCheck %s --check-prefixes=CHECK,CHECK-FP16
```

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp [hunk 7]
```diff
@@ -880,6 +880,8 @@ AArch64TargetLowering::AArch64TargetLowering(const TargetMachine &TM,
       setOperationAction(Op, MVT::f16, Legal);
   }
   // clang-format on
+  if (!Subtarget->hasFullFP16())
+    setOperationPromotedToType(ISD::FCANONICALIZE, MVT::f16, MVT::f32);
```
Comment (davemgreen) [inline llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:884]: Can this be added to LegalizeNarrowFP? I think that f32 is the default promotion type for f16.

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp [hunk 8]
```diff
@@ -1362,6 +1364,10 @@ AArch64TargetLowering::AArch64TargetLowering(const TargetMachine &TM,
         for (MVT Ty : {MVT::v4f16, MVT::v8f16})
           setOperationAction(Op, Ty, Legal);
     }
+    if (!Subtarget->hasFullFP16()) {
+      setOperationPromotedToType(ISD::FCANONICALIZE, MVT::v4f16, MVT::v4f32);
```
Comment (davemgreen) [inline llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:1368]: Can these be moved close to the existing setOperationPromotedToType around line 810?

**Review by davemgreen — APPROVED** [review]
Thanks for doing this. LGTM with some suggestions for where code can be moved.
